Parameterisation of secondary ice formation

oyvindseland opened the pull request:
Summary: Added new and extended parameterisation of secondary ice formation.

Contributors: @oyvindseland 

Reviewers: @gold2718 

Purpose of changes:   #152  Add parameteriation of secondary ice

Github PR URL: https://github.com/NorESMhub/CAM/pull/165

Changes made to build system: None

Changes made to the namelist: 5 new namelist parameters
- `rafsip_on`: If .true., compute secondary ice production using random forests method, Default: .false.
- `forestfileALL`: ML parameters for the forestALL RFR model, Default: None
- `forestfileBRDS`: ML parameters for the forestBRDS RFR model., Default: None
- `forestfileBRHM`: ML parameters for the forestBRHM RFR model., Default: None
- `forestfileBR`: ML parameters for the forestBR RFR model., Default: None
- `forestfileBRwarm`: ML parameters for the forestBRwarm RFR model., Default: None

Changes to the defaults for the boundary datasets: 5 ASCII files for reading in AI results used for calculating secondary ice formation.

Substantial timing or memory changes: None

Code changes to micro_mg2.0.F90 and added new subroutine module_random_forests.F90
Parameterisation is independent of Oslo-Aero and will in principle also work with MAM4.

1) New module contains random forest regressors for calculations of secondary ice formation for different physical domains (e.g. temperature ranges)
2) The main microphysics module is modified to include the new fluxes; the old limited parameterisation removed ; ice particle delimiter takes into account the new fluxes
3) The parameterisation is governed by  a logical rafsip_on. It is hardcoded and set to .true. Should be included as a namelist option.
4) The name of the files are hardcoded into the routine. The commit also include a  subroutine sec_ice_readnl that was used for reading in the file names from CAM namelist in project Forces simulations.

Georgakaki, P., &amp; Nenes, A. (2024). RaFSIP: Parameterizing ice multiplication in models using a machine learning approach. Journal of Advances in Modeling Earth Systems, 16, e2023MS003923.
https://doi.org/10.1029/2023MS003923

Only tested with compset NF2000climo at Betzy. Tested that the logical parameter set to .false. using the same model set-up give bit-identical results to model code built directly from the noresm2_3_develop branch

Issues addressed by this PR: Completes #152 for noresm2_3_develop

closes #152

Commit:
Added necessary changes for secondary ice formation

Commit:
Turned hardcoded logical parameter for parameterisation on

Commit:
File open statement made consistent for all new filew

gold2718 commented:
@oyvindseland, what is supposed to happen if `rafsip_on = .true.` but `do_cldice = .false.`?

oyvindseland commented:
I need to check it in more details but I thought that they should be independent of each other? Secondary ice formation does not require aerosols.

gold2718 commented:
> I need to check it in more details but I thought that they should be independent of each other? Secondary ice formation does not require aerosols.

Thanks, I just want to make sure the code behaves that way.

oyvindseland commented:
It is not critical that it works without do_cldice, but more general is always nice

gold2718 commented on src/NorESM/micro_mg2_0.F90:
```diff
@@ -1636,6 +1770,240 @@ subroutine micro_mg_tend ( &
 
+
+
+        IF (rafsip_on) THEN
```
This block is only executed if `do_cldice == .true.`. Is that intentional?

gold2718 commented on src/NorESM/micro_mg2_0.F90:
Can we get rid of `#ifndef NORESM2_ICE_DELIMITER` on line 2529, the `#else` clause beginning on line 2552. and the associated comments?

oyvindseland replied:
I need to check the code structure a bit since I realized that I did not fully understand the negation (infdef)
Does it mean ifndef  NORESM2_ICE_DELIMITER as if not ice delimiter as in noresm2, i.e. use the corrected version?

gold2718 replied:
This logic is from the NorESM2.1 code. By default, `NORESM2_ICE_DELIMITER` is _not_ set. You have to uncomment out the line above (`#define NORESM2_ICE_DELIMITER`) to get back to the NorESM2 behavior. We added this logic to NorESM2.1 because the model was more in balance that way. However, because the plan is to tune NorESM2.3 (at least a bit), I do not think we need this logic (to allow the NorESM2.0 behavior) anymore. Thoughts?

oyvindseland replied:
I agree that the test should be removed.

gold2718 reviewed (commented):
I have a question about removing an ifdef.

oyvindseland commented:
I finally got around to check the do_cldice and I see now that it is actually a check if MG parameterisation should be used, not only hetfreez, which have a different logical variable. 
--> rafsip_on should only be included if do_cldice == true . 

It might not be required but I have no idea what would happen then.

gold2718 commented:
@oyvindseland: We need a test for this scheme. What NorESM2.3 compset do you suggest?

oyvindseland commented:
It does not really matter but if we are going to tune the model N1850 is as good as anything else. The impact is more or less the same for both present dat and pre-industrial climate.

Commit:
Added namelist variables
Cleaned up calling of readnl and init routines
Cleaned up RaFSIP logic in micro_mg2_0.F90
Parallize initial data read
General code cleanup
Error if rafsip_on=.true. but micro_mg_do_cldice=.false.
Add test for secondary ice
BFB with NorESMhub/noresm2_3_develop when rafsip_on==.false.

oyvindseland committed:
Merge pull request #4 from gold2718/noresm2_3_secice

Added namelist variables and test

Commit:
Remove NorESM2.0 ice delimiter compile-time option

gold2718 committed:
Merge pull request #5 from gold2718/noresm2_3_secice

Remove NorESM2.0 ice delimiter compile-time option

BFB, checked against baseline

gold2718 reviewed (approved):
Looks good now.
